### PR TITLE
resolve the config file from the sudo user

### PR DESCRIPTION
**Pull request by runcom**
Sometimes docker is invoked directly with `sudo` (usually when docker is run from inside another binary). However it's common to have an unprivileged user in the docker group - and so the home config file is under `/home/unprivilegeduser/.docker/config`. This is usually fine because we don't need to use `sudo` with our user inside the docker group. When we call a binary with `sudo` which call docker itself the config file will be resolved to `/root/.docker` which is ofc empty (not exists usually). It would be good to support retrieving the calling sudo user and set the config dir to her home. A use case would be when we are authenticated against some registry as our unprivileged user and the docker pull/push/search command fails when called with `sudo` (because docker is looking for credentials inside `/root/.docker/`.

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
What should happen if a config file *does* exist in `/root/.docker`? Should we try that first, and if it *doesn't* exist fall back to `/unprivilegeduser/.docker`?

**Comment by runcom** [comment]
we can implement that check - stat `/root/.docker/`, if it doesn't exists, go on and check sudo user - does this work?

**Comment by thaJeztah** [comment]
Sounds good (I think). In general, I like the idea; on the other hand, we tell people to add users to a "Docker" group to prevent having to run sudo; https://github.com/docker/docker/blob/eb729a0cb0be1f2d2ba5c616fcd4e585289f6dfc/docs/installation/binaries.md#giving-non-root-access


Let me ping @tianon as well, who may have thought on what the "expected" approach would be

**Comment by runcom** [comment]
> Sounds good (I think). In general, I like the idea; on the other hand, we tell people to add users to a "Docker" group to prevent having to run sudo;

that's totally fine I guess - this would be just needed to ease the calling sudo processes in setting this correctly w/o having to do this manually

**Comment by runcom** [comment]
Updated to implement what @thaJeztah suggested

**Comment by rhatdan** [comment]
Until authentication is merged, it is more dangerous to use the docker group then to use sudo.  Also since we don't have RBAC controls on docker, it would be recommended to use sudo for sticking users into a particular container, rather or a set of commands rather then allowing them full access to the docker.socket.

http://www.projectatomic.io/blog/2014/09/granting-rights-to-users-to-use-docker-in-fedora/

**Comment by runcom** [comment]
tests seems ok, this change should be transparent to anyone using docker (with sudo also)

**Comment by tianon** [comment]
Seems sane to me. :+1:

**Comment by thaJeztah** [comment]
Alright, moving to code review

File: cliconfig/config.go [hunk 1]
```diff
@@ -178,7 +189,7 @@ func Load(configDir string) (*ConfigFile, error) {
 	}
 
 	// Can't find latest config file so check for the old one
-	confFile := filepath.Join(homedir.Get(), oldConfigfile)
+	confFile := getDefaultConfigDir(oldConfigfile)
```
Comment (thaJeztah) [inline cliconfig/config.go]: Perhaps we should keep this one as it was; wondering if we should remove this fallback completely (which version did we change to the new location?)

Note: removing the fallback to the old file should probably be a separate PR

Comment (runcom) [inline cliconfig/config.go]: > Perhaps we should keep this one as it was;

mmm this one also resolves the config from the home dir, why don't we want to change it?

Comment (thaJeztah) [inline cliconfig/config.go]: @runcom I was thinking to completely deprecate the `.dockercfg` file, so to not change the behavior for that one. Just searched when we moved to the new `.docker` directory, and that was in https://github.com/docker/docker/pull/12009, which was introduced in docker 1.7.0

Comment (runcom) [inline cliconfig/config.go]: make sense, I could do that in a followup PR

Comment (thaJeztah) [inline cliconfig/config.go]: Thanks!

**Comment by thaJeztah** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐮

**Comment by vdemeester** [comment]
It's green (so it's tempting to merge) but I think we could update a little bit the docs, maybe a *Note* in https://github.com/docker/docker/blob/master/docs/reference/commandline/cli.md :angel: 
/cc @thaJeztah

**Comment by thaJeztah** [comment]
Was just discussing with @vdemeester if docs changes are needed. We describe that the config file is "in your home directory"; https://github.com/docker/docker/blob/master/docs/reference/commandline/cli.md

So perhaps we don't need changes in the docs?

**Comment by runcom** [comment]
@thaJeztah we may add a note saying that `sudo` will correctly resolves the home directory of the calling user if the root directory is not found

**Comment by thaJeztah** [comment]
Alright, let's do that!

**Comment by runcom** [comment]
@thaJeztah @vdemeester updated

File: docs/reference/commandline/cli.md [hunk 2]
```diff
@@ -78,6 +78,11 @@ For example:
 Instructs Docker to use the configuration files in your `~/testconfigs/`
 directory when running the `ps` command.
 
+> **Note:**
+> Calling docker with `sudo` results in Docker looking first in `/root/.docker/`.
+> If it isn't found it looks for the user who did the `sudo` call and resolves the
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: Maybe we can simplify to this; (let me know if you think it's better)

```
> **Note**: If you run docker commands with `sudo`, Docker first looks for a configuration
> file in `/root/.docker/`, before looking in `~/.docker/` for the current user.
```

Comment (runcom) [inline docs/reference/commandline/cli.md]: mmm `for the current user` is wrong because it specifically resolves for the current *sudo* user

Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: Yes, I doubted about that one; OTOH, wondering if most users will consider "the user that did the sudo call" the "current" user? i.e. if I'm logged in as "sebastiaan", and run `sudo foo`, I can imagine most people will consider "sebastiaan" the current user. But open to suggestions

Comment (runcom) [inline docs/reference/commandline/cli.md]: I would say the opposite :/ sudo something as my runcom user will be executed with root :/ btw, I don't have a strong opinion :)

Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: Perhaps "before looking in `~/.docker/` for the user who did the `sudo` call", or "the logged-in user" (would logged-in be less confusing?)

Comment (vdemeester) [inline docs/reference/commandline/cli.md]: Hum, I like the *"the user that did the sudo call"* in fact because it's not prone to interpretation. At first, current user for me means the one that did the sudo call, but I can see some confusion basde on your comment @runcom. So I'm fine with the current sentences. :stuck_out_tongue_closed_eyes:

Comment (runcom) [inline docs/reference/commandline/cli.md]: I'll change it to "the user who did the sudo call" :)

Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: The original version, or my version with "the user that did the `sudo` call" at the end?

**Comment by runcom** [comment]
@thaJeztah @vdemeester updated

**Comment by thaJeztah** [comment]
awesome, LGTM!

**Comment by runcom** [comment]
Ping @vdemeester (tests seems flaky)

**Comment by vdemeester** [comment]
LGTM 🦁

**Comment by runcom** [comment]
flaky ci is flaky :(

**Comment by thaJeztah** [comment]
Reverted in https://github.com/docker/docker/pull/20780, so removing impact labels
